### DISCUSSION TITLE
Client extensions, decimal, serialization and RSC

### DISCUSSION THREAD

**Discussion by vimtor** [prisma/prisma#19983] [Q&A] [answered]
When using Prisma client extensions with React server components, you may see warnings like these: 

```
Warning: Only plain objects can be passed to Client Components from Server Components. Objects with symbol properties like nodejs.util.inspect.custom are not supported.
```

```
Warning: Only plain objects can be passed to Client Components from Server Components. Decimal objects are not supported.
```

I understand the issue, but I'm not sure what the optimal solution is. Mapping the whole object doesn't seem ideal. For now, it's just a warning, but I'm worried it could become an error in the future.

I brought this up in the context of RSC, but it also happens when using something like `Decimal`. Since this is an unsolved/unplanned issue I would like to open up the discussion to see how we could be able to solve it.

What solutions have you found? What ideas do you have in mind?

Since the Prisma team has a lot of experience with Rust, maybe the creation of a custom SWC plugin could be a great solution for Next.js users. [Similar to what Blitz.js is doing with Superjson](https://github.com/blitz-js/next-superjson-plugin).

### Related issues and discussions:

- https://github.com/vercel/next.js/issues/47447
- https://github.com/prisma/prisma/issues/9170
- https://github.com/prisma/prisma-examples/issues/3047
- https://github.com/prisma/prisma/issues/6049

**Comment by ludralph** [comment, marked as answer]
Hi @vimtor 👋 

Thank you for opening this discussion. One approach you mentioned is manually serializing the objects with non-plain properties into plain objects before passing them from server to client components. However, as you said, there may be better approaches. Creating a custom SWC plugin could be a potential solution, and this would require some development effort. I have also shared this internally with the engineering team for their thoughts.

**Reply by ludralph** [reply]
Hi @vimtor 

The advice from the team at this point is to use [devalue](https://github.com/Rich-Harris/devalue) or [superjson](https://github.com/blitz-js/superjson) for serializing what is not serializable with JSON for now. We also discussed at some point if we want to have built-in `serializeResponse` / `deserializeResponse` functions in Prisma, but at the moment, there is no strong need for this; community libraries work fine. Regarding the symbol warning: it’s fine to ignore for now. We’ll see if we can hide it better. Symbol key is there not for any production-critical functionality, just for nicer console.log while debugging. In the worst case, we can just remove it if it becomes a problem

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by j0nas** [reply]
Hi @ludralph. I've pretty much spent the entire day scouring the internet for how to solve this issue in a non-hacky way, without being able to find much.

```
 Warning: Only plain objects can be passed to Client Components from Server Components. Decimal objects are not supported.
   {id: 9, createdAt: Date, updatedAt: ..., imageId: ..., name: ..., price: Decimal, season: ..., addressId: ..., photographerId: ..., image: ..., address: ..., tags: ...}
                                                                            ^^^^^^^
```

I have a model with a `Decimal` field and I want to pass it from a server component to a client component in Next.js. Is the recommended way to serialize the object "manually"? That's not very elegant; it requires wrapping objects on a per-case basis. I've also tried to extend the Prisma client, like so: 

```
  ...
  }).$extends({
    result: {
      areaPicture: {
        price: {
          compute({ price }: { price: Decimal }) {
            return price.toString();
          },
        },
      },
    },
  });
```

However, that simply gives me another warning: 
```
[app] Warning: Only plain objects can be passed to Client Components from Server Components. Objects with symbol properties like nodejs.util.inspect.custom are not supported.
```

It would be super neat to add a few lines in the Prisma docs about what recommended/possible solutions to this is. 🙏

**Reply by divvjson** [reply]
@j0nas I have this exact issue. How did you ultimately choose to handle this?

**Comment by fabioluiz1** [comment]
This is a really annoying behavior. Given Decimal is a type supported at the ORM level, it should have proper serialization solution built in on Prisma. 

Using Next.js 14 and server actions, I have been manually serializing decimal attributes before sending data to client components. However, the worst scenario is when you add a new Decimal column into an existing model that is loaded into many existing components, because they are reused in many server actions: adding the decimal column becomes a real nightmare because all the existing components will print warnings into the console.

Bottom line - a simple task of adding a decimal column has a big impact, demanding to change many (sometimes dozens of) components through the entire app.

**Comment by omararturo** [comment]
Any solution? Date is an object and is supported by NextJS, why Decimal, supported for many ORM like prisma. Manual format is a bad and no scalable solution.

**Comment by martin-purplefish** [comment]
Just wanted to chime in and say that this is really annoying for us as well. It feels like there should be a way to have custom serializers on types like this.

**Comment by compnasajedalen** [comment]
Any solution?

**Reply by ndanyikevin** [reply]
I'm also spying for a solution. Nextjs throwing 33 errors

**Comment by goldengecko** [comment]
This one caught me too - I'm building a nuxt app and my decimals are arriving as strings. I have many decimals in my database, so picking out all the fields and handling them manually really isn't an option. 

It seems that I should in theory be able to use prisma.$extends to be able to process the fields but so far I haven't been able to work out how to from the documentation.

**Reply by goldengecko** [reply]
I'm kind of thinking that it should be along the lines of this:


const prisma = new PrismaClient().$extends({
    query: {
      $allModels: {
        $allOperations({ model, operation, args, query }) {
          // Process results for queries
          return query(args).then((result) => {
            if (
              operation === 'findMany' ||
              operation === 'findUnique' ||
              operation === 'findFirst'
            ) {
              return processDecimals(result, model)
            }
            return result
          })
        },
      },
    },
  })

**Comment by goldengecko** [comment]
OK, I looked into this more, and the core issue is that it retrieves the data in the right format (using Prisma.Decimal) on the server side, but when it is serialised to JSON to send to the client, the Decimal fields are converted to strings. This is to ensure that no precision gets lost along the way.

So there are a few options:

1. Leave it as is and on the client side understand which fields were supposed to be Decimals and convert/process accordingly.
2. Convert the Decimals to Numbers during the JSON serialisation. While this means that consuming the API JSON is easy, it has the potential downside of losing precision.
3. Convert the Decimals to some other format and convert them back on the client side. This has the downside that if anyone else wants to consume the API output they would need to know how to convert it back, but it does not lose precision.

For my needs I decided to go with option 3.

In order to do this, server side, I added this code:


```typescript
import { Prisma } from '@prisma/client'

(Prisma.Decimal.prototype as any).toJSON = function () {
  return { __decimal__: this.toString() }
}

```

this means that for instance the JSON for my quantity Decimal field would come out something like this:
```json
"quantity": {
    "__decimal__": "400"
},
```

Now when I consume the API response I have a function to convert it back:
```typescript
import { Prisma } from '@prisma/client'

function decimalReviver(key: string, value: any) {
  if (value && typeof value === 'object' && '__decimal__' in value) {
    // Convert back into a Decimal (or decimal.js if you prefer).
    return new Prisma.Decimal(value.__decimal__)
  }
  return value
}
```

Since I am working in Nuxt 3, both useFetch and $fetch accept a parameter called `parseResponse` which allows me to automatically do the conversion of the response like this:
```typescript
const { data, status } = useFetch('/api/wherever', { otherArgs, parseResponse: (text) => JSON.parse(text, decimalReviver), })
```

And I never have to think about it again.

Clearly if you wanted to go with option 2 above, all you would need to do is the first of these steps and override toJSON to return a number instead of a custom object.

Hope this helps someone.

**Comment by RonAlmog** [comment]
Yup, this is a bummer. Prisma, please find a solution. we need our decimals...

**Comment by danielnathanael** [comment]
I managed to solve this warning by using serialize / deserialize function from [SuperJSON package](https://www.npmjs.com/package/superjson)

**Reply by hjoelh** [reply]
Could you elaborate please?

**Comment by ScreamZ** [comment]
I was thinking about implementing https://github.com/flightcontrolhq/superjson#decimaljs--prismadecimal with superjson and babel plugin

**Comment by HassanZohaib121** [comment]
add a generic prisma wrapper so all the quries will be converted to the plain object.

import { PrismaClient, Prisma } from "@prisma/client";

const prisma = new PrismaClient();

/**
 * Recursively converts Prisma.Decimals → number and Date → string
 */
function serializePrisma<T>(obj: T): T {
  if (obj === null || obj === undefined) return obj;

  if (obj instanceof Prisma.Decimal) return obj.toNumber() as unknown as T;
  if (obj instanceof Date) return obj.toISOString() as unknown as T;

  if (Array.isArray(obj)) return obj.map(serializePrisma) as unknown as T;

  if (typeof obj === "object") {
    const result: Record<string, any> = {};
    for (const key in obj) {
      if (Object.hasOwn(obj, key)) {
        result[key] = serializePrisma((obj as any)[key]);
      }
    }
    return result as T;
  }

  return obj;
}

/**
 * Wrapper for findMany, findFirst, findUnique, etc.
 * Automatically serializes all Prisma results.
 */
export async function findManySerialized<T>(model: keyof PrismaClient, args?: any) {
  const data = await (prisma[model] as any).findMany(args);
  return serializePrisma(data);
}

export async function findFirstSerialized<T>(model: keyof PrismaClient, args?: any) {
  const data = await (prisma[model] as any).findFirst(args);
  return serializePrisma(data);
}

export async function findUniqueSerialized<T>(model: keyof PrismaClient, args?: any) {
  const data = await (prisma[model] as any).findUnique(args);
  return serializePrisma(data);
}

// Export the original Prisma client in case you need it
export { prisma, serializePrisma };
